File: Implementations/my_stack.py
from my_array import Array

class Node():
    def __init__(self, value):
        self.value = value
        self.next = None
        
class StackLL():
    def __init__(self):
        self.top = None
        self.bottom = None
        self.length = 0
        
    def peek(self):
        return self.top
        
    def push(self, value):
        new = Node(value)
        if self.length == 0:
            self.top, self.bottom = new, new
        else:
            new.next = self.top
            self.top = new
        self.length += 1
        return self
        
    def pop(self):
        if self.top != None:
            self.top = self.top.next
            stack.length -= 1
        return self


class StackA():
    def __init__(self):
        self.data = Array()  
        self.length = 0  

    def peek(self):
        if self.length == 0:
            return None 
        return self.data.get(self.length - 1) 

    def push(self, value):
        self.data.append(value)  
        self.length += 1
        return self

    def pop(self):
        if self.length > 0:
            self.data.pop() 
            self.length -= 1
        return self

            
stack = StackA()
stack.push(5)
stack.push(15)
stack.pop()
for i in range(5):
	stack.push(i**2//2)
print(stack.peek())
print(vars(stack.data))

File: cdsa/my_stack.c
#include "my_stack.h"
#include <stdlib.h>

static int _ll_peek(StackLL *self) { return self->top ? self->top->value : -1; }

static void _ll_push(StackLL *self, int value) {
    StackNode *node = (StackNode *)malloc(sizeof(StackNode));
    node->value = value;
    node->next  = NULL;
    if (self->length == 0) { self->top = node; self->bottom = node; }
    else { node->next = self->top; self->top = node; }
    self->length++;
}

static void _ll_pop(StackLL *self) {
    if (!self->top) return;
    StackNode *old = self->top;
    self->top = self->top->next;
    free(old);
    self->length--;
}

static void _ll_destroy(StackLL *self) {
    StackNode *cur = self->top;
    while (cur) { StackNode *next = cur->next; free(cur); cur = next; }
    free(self);
}

StackLL *StackLL_new() {
    StackLL *s  = (StackLL *)malloc(sizeof(StackLL));
    s->top      = NULL;
    s->bottom   = NULL;
    s->length   = 0;
    s->peek     = _ll_peek;
    s->push     = _ll_push;
    s->pop      = _ll_pop;
    s->destroy  = _ll_destroy;
    return s;
}

static int _a_peek(StackA *self) {
    return self->length ? self->data->get(self->data, self->length - 1) : -1;
}

static void _a_push(StackA *self, int value) {
    self->data->append(self->data, value);
    self->length++;
}

static void _a_pop(StackA *self) {
    if (self->length > 0) { self->data->pop(self->data); self->length--; }
}

static void _a_destroy(StackA *self) {
    self->data->destroy(self->data);
    free(self);
}

StackA *StackA_new() {
    StackA *s  = (StackA *)malloc(sizeof(StackA));
    s->data    = Array_new();
    s->length  = 0;
    s->peek    = _a_peek;
    s->push    = _a_push;
    s->pop     = _a_pop;
    s->destroy = _a_destroy;
    return s;
}
